### PR TITLE
Add Fenix feature annotations, based on issue metadata

### PR DESCRIPTION
**Pull request by wlach**
In many cases, is possible to infer the feature a fenix metric corresponds to from the issue metadata corresponding to the metric landing. This is an experiment in doing exactly this, via some automated scripts (stored in the `scripts/` subdirectory)

Things to explore before landing:

* [x] What percentage of Fenix metrics does this actually cover? It excludes libraries Fenix uses (naturally) ~~but I also noticed some weirdly or incorrectly specified issues that my parser didn't catch (will file an issue in fenix itself to fix those)~~ (update: these were old metrics that are no longer used or in the source code). Also some issues appear to only specify the "Telemetry" feature ([example](https://github.com/mozilla-mobile/fenix/issues/6834)), which we just skip over since that provides no useful information.
* [x] What should the view of this in the Glean Dictionary be? Obvious things would be adding these labels in the metrics list (and letting people filter on them) as well as making it more clear in the metric description as well.
* [x] Do we want to define some kind of application-level schema of what the allowed values for these "features" should be?
* [x] How do we keep these up to date with new annotations? The scripts currently assume you're doing this from scratch.
* [ ] ... maybe other things I can't think of ...

### CODE REVIEW

**Comment by wlach** [comment]
@jklukas When you have a minute, I wouldn't mind your feedback on this approach. I'm also going to see if the Fenix folks have anything to say about this.

**Comment by jklukas** [comment]
Taking a look now. Apologies that this got buried in my inbox.

File: scripts/create-api.py [hunk 1]
```diff
@@ -1,36 +1,56 @@
...
+        metadata = yaml.load(open(os.path.join(app_dir, "metadata.yaml")))
+        data[app] = metadata
+        valid_components = metadata.get("components", []).keys()
+    except:
+        data[app] = {}
+        pass
+    data[app]["annotations"] = defaultdict(lambda: {})
     for annotation_type in ("metrics", "pings"):
-        annotation_dir = os.path.join(ANNOTATIONS_DIR, app, annotation_type)
+        annotation_dir = os.path.join(app_dir, annotation_type)
```
Comment (jklukas) [inline scripts/create-api.py:28]: Using `pathlib.Path` objects can clean a good bit of this up, avoiding the need for `os.path.join` calls in favor of `app_dir / annotation_type`, `os.listdir` calls can be replaced with `p.iterdir()`, etc.

File: scripts/create-api.py [hunk 2]
```diff
@@ -1,36 +1,56 @@
...
+            components = annotation_md.get("components")
+            if components:
+                invalid_components = [
+                    component
+                    for component in components
+                    if component not in valid_components
+                ]
+                if invalid_components:
+                    sys.stderr.write(
+                        f"Invalid components found in {annotation_filename}: {invalid_components}"
```
Comment (jklukas) [inline scripts/create-api.py:49]:
```suggestion
                        f"Invalid components found in {annotation_filename}: {invalid_components};"
                        f" if these should be accepted values, update annotations/{app}/metadata.yaml"
```

File: scripts/create-feature-annotations.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
+#!/usr/bin/env python
```
Comment (jklukas) [inline scripts/create-feature-annotations.py:1]: For executable scripts, I recommend naming in kebab-case (which you're doing here) with no extension. The language shouldn't be part of the user interface for a CLI script.

File: scripts/create-feature-annotations.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
```
Comment (jklukas) [inline scripts/create-feature-annotations.py:2]: There's not enough documentation here yet for me to quickly understand the purpose of each script and how they're intended to be used.

File: scripts/scrape-fenix.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import sys
+
+import requests
+
+
+FENIX_METRICS_URL = 'https://probeinfo.telemetry.mozilla.org/glean/fenix/metrics'
+GITHUB_TOKEN = os.getenv('GITHUB_TOKEN')
+
+if not GITHUB_TOKEN:
+    print("Github token required!")
```
Comment (jklukas) [inline scripts/scrape-fenix.py:15]:
```suggestion
    print("Github token required! Please set the GITHUB_TOKEN environment variable.")
```

File: requirements.txt [hunk 6]
```diff
@@ -1 +1,2 @@
-python-frontmatter
\ No newline at end of file
+python-frontmatter
+PyYAML
```
Comment (jklukas) [inline requirements.txt:2]:
```suggestion
PyYAML
requests
```

File: scripts/create-feature-annotations.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+feature_map = json.load(open(sys.argv[1]))
+for metric, features in feature_map.items():
+    features = set([feature.replace("Feature:", "") for feature in features])
+    try:
+        # the telemetry feature is not what we're tracking
+        features.remove("Telemetry")
+    except KeyError:
+        pass
+    if len(features):
+        metric_path = os.path.join("annotations", "fenix", "metrics", metric)
```
Comment (jklukas) [inline scripts/create-feature-annotations.py:16]: Note that this script is fenix-specific, but the name of the script doesn't make that clear.

**Review by jklukas — COMMENTED** [review]
This looks reasonable so far. Let me know when this is ready for more focused code review.

File: scripts/create-api.py [hunk 8]
```diff
@@ -33,36 +36,55 @@ def linkify(text):
...
+        annotation["labels"] = labels
+
     return annotation
 
 
 for app in apps:
     try:
         data[app]["app"] = read_annotation(
-            os.path.join(ANNOTATIONS_DIR, app, "README.md")
+            os.path.join(ANNOTATIONS_DIR, app, "README.md"), []
```
Comment (wlach) [inline scripts/create-api.py:64]: I updated the code in my scraper script to use Pathlib as @jklukas suggested, but modifying this one to do the same seems out of scope. Let's handle that in a followup.

File: scripts/scrape-fenix [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+
+Usage:
+
+./scripts/update_metric_labels
+
+This is intended as a stopgap until we can incorporate this information directly
+into Fenix itself. See:
+
+https://docs.google.com/document/d/1K-Hl_ZWRvVzqBAGzdXb_oQY52D-D5YDxorYmzyzfxuw/edit#
+"""
```
Comment (wlach) [inline scripts/scrape-fenix:15]: I consolidated all the miscellaneous scraper logic into this single script, which will hopefully be easier to reason about.

**Comment by wlach** [comment]
I think this is ready for another look. This is intended more as an initial attempt than a long term piece of infrastructure. In particular, I'm hoping the need for scraping will be temporary.

File: scripts/scrape-fenix [hunk 10]
```diff
@@ -0,0 +1,179 @@
+#!/usr/bin/env python
+
+"""
+Toolkit for scraping GitHub feature labels from Fenix data reviews and using
+them to create a preliminary set of metric annotations.
+
+Usage:
+
+./scripts/update_metric_labels
```
Comment (jklukas) [inline scripts/scrape-fenix:9]:
```suggestion
./scripts/scrape-fenix
```

**Review by jklukas — APPROVED** [review]

**Pull request merged**